python/pyiceberg/transforms.py

```diff
@@ -249,6 +294,20 @@ def satisfies_order_of(self, other: Transform) -> bool:
     def result_type(self, source: IcebergType) -> IcebergType:
         return IntegerType()
 
+    def project(self, name: str, pred: BoundPredicate) -> Optional[UnboundPredicate]:
```

Comment:
I think the date/time transforms need to use the same truncation logic as the `truncate` function. That's because the operation may need to change. For example: `ts < '2022-11-06T15:56'` with a `day` transform should be `ts_day <= '2022-11-06'`.

Reply:
Nice, thanks! That was one of the areas I was unsure of. Since we can simplify this quite a lot on the Python side, it was a bit tricky to translate. I also added tests:
```python
def test_projection_day_human(bound_reference_date: BoundReference) -> None:
    date_literal = literal(date(2018, 1, 1))
    assert DayTransform().project("dt", BoundEqualTo(term=bound_reference_date, literal=date_literal)) == EqualTo(
        term="dt", literal=literal(17532)
    )  # == 2018, 1, 1

    assert DayTransform().project("dt", BoundLessThanOrEqual(term=bound_reference_date, literal=date_literal)) == LessThanOrEqual(
        term="dt", literal=literal(17532)
    )  # <= 2018, 1, 1

    assert DayTransform().project("dt", BoundLessThan(term=bound_reference_date, literal=date_literal)) == LessThanOrEqual(
        term="dt", literal=literal(17531)
    )  # <= 2017, 12, 31

    assert DayTransform().project(
        "dt", BoundGreaterThanOrEqual(term=bound_reference_date, literal=date_literal)
    ) == GreaterThanOrEqual(
        term="dt", literal=literal(17532)
    )  # >= 2018, 1, 1

    assert DayTransform().project("dt", BoundGreaterThan(term=bound_reference_date, literal=date_literal)) == GreaterThanOrEqual(
        term="dt", literal=literal(17533)
    )  # >= 2018, 1, 2
```

Reply:
Looks correct now.